#!/usr/share/ucs-test/runner python
## desc: Create DHCP subnet, check it's attributes and remove it again
## tags: [udm-dhcp]
## roles: [domaincontroller_master]
## exposure: careful
## packages:
##   - univention-config
##   - univention-directory-manager-tools



import sys
import univention.testing.udm as udm_test
import univention.testing.strings as uts
import univention.uldap as uldap

if __name__ == '__main__':
	with udm_test.UCSTestUDM() as udm:
		dhcpService = udm.create_object('dhcp/service', service = uts.random_name())
		dhcpSubnetProperties = {
			'subnet': '10.20.30.0',
			'subnetmask': '24',
			'broadcastaddress': '10.20.30.255',
			'range': '10.20.30.1 10.20.30.254'
		}
		dhcpSubnet = udm.create_object('dhcp/subnet', superordinate = dhcpService, **dhcpSubnetProperties)

		ldapMapping = {'subnet':'cn', 'subnetmask':'dhcpNetMask', 'broadcastaddress':'univentionDhcpBroadcastAddress', 'range':'dhcpRange'}
		ldap = uldap.getAdminConnection()
		ldapFilter = '(&(cn=10.20.30.0)(objectClass=univentionDhcpSubnet))'
		for dn, attr in ldap.search(base = dhcpService, filter = ldapFilter, attr = ldapMapping.values()):
			if dn == dhcpSubnet:
				for attribute, value in dhcpSubnetProperties.items():
					if [value] != attr[ ldapMapping[attribute] ]:
						print 'Test FAILED. Expected "%s" of created DHCP subnet to be %r, but %r was found instead.' % (ldapMapping[attribute], [value], attr[ldapMapping[attribute]])
						sys.exit(1)
				break
		else:
			print 'Test FAILED. Could not find created DHCP subnet "%s" in LDAP.' % dhcpSubnet
			sys.exit(1)

		udm.remove_object('dhcp/subnet', dn = dhcpSubnet, superordinate = dhcpService)
		if dhcpSubnet in ldap.searchDn(base = dhcpService, filter = ldapFilter):
			print 'Test FAILED. Could still find created DHCP subnet "%s" in LDAP after removing it.' % dhcpSubnet
			sys.exit(1)


